kv/util/runtime/value.go

```diff
@@ -141,49 +148,75 @@ func (v *value) Unwatch() {
 	if v.status == valueNotWatching {
 		return
 	}
-	v.watch.Close()
+	v.updatable.Close()
 	v.status = valueNotWatching
-	v.watch = nil
+	v.updatable = nil
 }
 
-func (v *value) watchUpdates(watch kv.ValueWatch) {
-	for range watch.C() {
+func (v *value) watchUpdates(updatable Updatable) {
+	for range updatable.C() {
 		v.Lock()
 		// If we are not watching, or we are watching with a different
 		// watch because we stopped the current watch and started a new
 		// one, return immediately.
-		if v.status != valueWatching || v.watch != watch {
+		if v.status != valueWatching || v.updatable != updatable {
 			v.Unlock()
 			return
 		}
-		if err := v.updateWithLockFn(watch.Get()); err != nil {
+		update, err := v.getFn(updatable)
+		if err != nil {
+			v.log.Errorf("error getting update: %v", err)
+			v.Unlock()
+			continue
+		}
+		if err = v.processWithLockFn(update); err != nil {
 			v.log.Errorf("error updating value: %v", err)
 		}
 		v.Unlock()
 	}
 }
 
-func (v *value) updateWithLock(update kv.Value) error {
+func (v *value) processWithLock(update interface{}) error {
 	if update == nil {
 		return errNilValue
 	}
-	if v.currValue != nil && !update.IsNewer(v.currValue) {
-		v.log.Warnf("ignore kv update with version %d which is not newer than the version of the current value: %d", update.Version(), v.currValue.Version())
+	if !v.isNewerUpdate(update) {
 		return nil
 	}
-	v.log.Infof("received kv update with version %d for key %s", update.Version(), v.key)
-	latest, err := v.unmarshalFn(update)
-	if err != nil {
-		err = fmt.Errorf("error unmarshalling value for version %d: %v", update.Version(), err)
-		return err
-	}
-	if err := v.processFn(latest); err != nil {
+	if err := v.processFn(update); err != nil {
 		return err
 	}
 	v.currValue = update
 	return nil
 }
 
+func (v *value) isNewerUpdate(update interface{}) bool {
+	if v.currValue == nil {
+		return true
+	}
+	versionedCurrValue, ok := v.currValue.(kv.Versionable)
+	if !ok {
+		return true
+	}
+	versionedUpdate, ok := update.(kv.Versionable)
+	if !ok {
```

Comment:
This is kinda weird, can we just make the `update` always a `Versionable`?

Reply:
see comment above